### ISSUE TITLE
M-FSDP: reduce-scatter input buffers are held until the end of backward

### ISSUE THREAD

**Issue by conver334** [ISEEKYAN/Megatron-LM#158]
## Summary

In M-FSDP's backward pass, every bucket allocates an fp32 reduce-scatter input buffer sized to the bucket's **unsharded** `full_numel`, and all of them are launched before any of them is waited on. The buffers are therefore live simultaneously, so M-FSDP's transient working set is **O(number of layers)**. FSDP2 keeps a constant number of RS buffers alive, so its working set is O(1) in layer count.

On Qwen3.5-35B-A3B at 16 GPUs this costs **+14.8 GiB of peak memory**: M-FSDP peaks at 58.1 GiB where `fsdp2` peaks at 43.3 GiB. Resident state is not the cause — the two are within 0.24 GiB of each other there.

## Environment

- `experimental/lite`, backend `mlite`, `impl_cfg.optimizer` in `{fsdp2, mfsdp}`
- Reproduced at `0d62b7f26` (fix of b1351ae34b534b710964044e3ef62ec0011ac82d)
- torch 2.10.0+cu129, TE 2.12.0+5671fd3, 16x H100 80GB (2 nodes)
- Model: Qwen3.5-35B-A3B (`model_name=qwen3_5`), TP2 / ETP1 / EP8 / CP1 / PP1
- `--seq-len 128 --num-microbatches 1 --seed 42`, no CPU offload, `use_precision_aware_optimizer=false`, `offload_fraction=0.0`
- `CUDA_DEVICE_MAX_CONNECTIONS=8` for every backend

## Reproduction

Sweep `--truncate-layers` with everything else fixed, for both backends:

```bash
cd experimental/lite
export PYTHONPATH=$PWD:/path/to/Megatron-LM
export MEGATRON_LITE_DETERMINISTIC=1 CUBLAS_WORKSPACE_CONFIG=:4096:8 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0

for L in 10 20 40; do
for OPT in fsdp2 mfsdp; do
  CUDA_DEVICE_MAX_CONNECTIONS=8 \
  torchrun --nnodes 2 --node_rank ${NODE_RANK} --master_addr ${MASTER_ADDR} \
      --master_port 29700 --nproc_per_node 8 \
    <driver> --backend mlite \
    --hf-path <Qwen3.5-35B-A3B> --model-name qwen3_5 --impl lite \
    --tp 2 --etp 1 --ep 8 --cp 1 --pp 1 --num-microbatches 1 \
    --steps 3 --seq-len 128 --seed 42 --same-data-across-dp --use-thd \
    --truncate-layers ${L} \
    --override-optimizer-json '{"use_precision_aware_optimizer":false,"offload_fraction":0.0}' \
    --impl-cfg-json "{\"optimizer\":\"${OPT}\",\"mount_vision_model\":false}"
done
done
```

Peak is read from `torch.cuda.max_memory_allocated()`. **Discard step 0**: `torch.optim.AdamW` creates `exp_avg`/`exp_avg_sq` on its first `step()`, so step 0's forward runs before the Adam state exists and its peak understates the steady state (on this config it reads 41.6 GiB vs a steady 58.1 GiB).

## Observed

Steady-state peak allocated (GiB), rank world-max:

| L | `fsdp2` peak | `mfsdp` peak | delta |
| ---: | ---: | ---: | ---: |
| 10 | 13.893 | 18.421 | +4.53 |
| 20 | 23.707 | 31.656 | +7.95 |
| 40 | 43.342 | 58.145 | **+14.80** |

Both are linear in `L`, but with different slopes — `fsdp2` 0.982 GiB/layer, `mfsdp` 1.324 GiB/layer. The gap grows without bound as the model gets deeper.

Splitting each peak into resident state (measured after a steady-state `optimizer.step()`, with gradients still live) and the transient excess during forward/backward:

| L | `fsdp2` resident | `fsdp2` transient | `mfsdp` resident | `mfsdp` transient |
| ---: | ---: | ---: | ---: | ---: |
| 10 | 11.287 | **2.606** | 14.523 | **3.898** |
| 20 | 21.100 | **2.606** | 23.337 | **8.319** |
| 40 | 40.736 | **2.606** | 40.973 | **17.171** |

`fsdp2`'s transient is bit-identical at all three L — O(1), as intended. M-FSDP's is linear.

> Note on the split: resident must be sampled at `post_opt`, not at the next step's `pre`. `fsdp2`'s `zero_grad` sets grads to `None` and drops ~8 GiB there, while M-FSDP keeps a persistent flat `main_grad_buffer`; anchoring on `pre` charges `fsdp2`'s gradients to the transient column and makes the two incomparable.

Resident state is **not** the problem: at L=40 the two agree within 0.24 GiB, both matching `20 B/param x 2.1676e9 = 40.37 GiB`. (M-FSDP is actually 2 B/param cheaper resident — fp32 master + fp32 main_grad + Adam m/v + a bf16 compute copy = 18 B/param — but pays ~4.24 GiB fixed for the root bucket and the arena, which cancels out at 40 layers.)

## The transient closes exactly against bucket geometry

Summing every bucket's `full_numel` (unsharded element count) and pricing it as fp32:

| L | buckets | sum `full_numel` | x4 B | `mfsdp` transient | residual |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 10 | 53 | 1,698,365,696 | 6.327 | 3.898 | **2.429** |
| 20 | 103 | 2,884,942,688 | 10.747 | 8.319 | **2.429** |
| 40 | 203 | 5,261,323,968 | 19.600 | 17.171 | **2.429** |

The residual is the same constant to the last digit at all three L (it is the part already counted in the persistent arena: the root buckets' fp32 grad slots, 2 x 254,279,680 x 4 B = 1.895 GiB, plus 0.534 GiB of other resident items).

The implied slope agrees independently. Per-layer `full_numel` on this config:

```
expert   39,845,888 x 2 + 20,971,520        = 100,663,296
dense   (30 x 18,958,336 + 10 x 15,730,688) / 40 =  18,151,424
per layer                                   = 118,814,720 elements
x 4 B (fp32) = 0.4426 GiB/layer
measured slope (17.171 - 8.319) / 20        = 0.4426 GiB/layer
```

bf16 would be 0.2213; any DP-sharded width would be 1/2 or 1/8 of that. **Only "fp32 x unsharded x one per layer" fits.**

## Root cause

**1. Each bucket leases a full-size fp32 buffer.** `ParamBucket.prepare_grad_reduce`, `mfsdp/buffer.py:583`:

```python
self._grad_lease = self.allocator.allocate(
    self.full_numel,                     # unsharded
    dtype=self.policy.grad_comm_dtype,   # defaults to torch.float32 (config.py:42)
    ...
)
```

It is released only at the end of `wait_grad_reduce` (`buffer.py:643-645`), so the lease lives from launch until that bucket is waited on.

**2. The scheduler waits on nothing until every bucket has launched.** `GradReducePipeline.finish`, `mfsdp/buffer.py:1051`:

```python
def finish(self) -> None:
    for bucket in reversed(self.buckets):
        self.reduce_gradients(bucket, force=True)   # allocate + launch, all N
    self.comm_stream.wait_for_current()
    for bucket in self.buckets:
        bucket.wait_grad_reduce()                   # only now released
```

Between the two loops, all N leases are live at once.

`overlap_grad_reduce` (default `True`, `config.py:37`) does not change this. It installs `reduce_gradients` as `grad_ready_callback` (`buffer.py:1028`) so each bucket launches as soon as its grads are ready during backward (`buffer.py:701-702`) — but nothing reclaims the lease when that bucket's reduce-scatter completes; the waits still all happen in `finish()`. Overlap only makes the leases live *longer*.

## Why the double-buffer pool does not bound this

`DoubleBufferAllocator` keeps 2 slots per layout key. All 40 layers share the same expert-bucket layout key, so they contend for the same 2 slots — and when both are busy, `allocate` falls through to the base allocator (`buffer.py:211-216`):

```python
for slot in range(2):
    if slot in busy:
        continue
    ...
    return BufferLease(..., slot=slot, ...)
return super().allocate(numel, dtype=dtype, device=device, group=group, key=pool_key)
```

That fallthrough is a plain `torch.empty` that is **not** tracked in `_slots` and is freed only when the lease drops its last reference — i.e. in `wait_grad_reduce`. So under the current scheduling, buckets 3..N each get a fresh full-size fp32 allocation off the caching allocator, and all of them coexist. The 2-slot pool stays at a constant 4.355 GiB across all three L in the measurements above, confirming the growth is entirely in the spilled allocations.

## Suggested fix

Bound the number of in-flight reduce-scatter leases: wait on and release bucket `k` before launching bucket `k+W`, with `W = 2` to match the double-buffer pool (and FSDP2's behaviour). Equivalently, reclaim each lease as soon as its collective completes rather than deferring all waits to `finish()`.

Expected on the config above (L=40):

```
transient  17.171 -> ~0.9 GiB     (about two layers' worth)
peak       58.145 -> ~41.9 GiB    (below fsdp2's 43.342 and dist_opt's 55.764)
mfsdp/fsdp2 peak ratio  1.342 -> ~0.97
```

The cost is that the reduce-scatters can no longer all be issued up front. Under EP that is not an additional loss: parameter-gather overlap is already force-disabled for any multidimensional composition by `_order_param_gathers_for_parallel_collectives` (`mfsdp/optimizer.py:474`, triggered whenever `tp*cp*ep*etp != 1` — EP=8 alone is enough, TP is not involved).
